// ngx-flip/esm5/lib/flip.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/flip.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, Input, Renderer2, ViewChild } from '@angular/core';
var FlipComponent = /** @class */ (function () {
    function FlipComponent(renderer) {
        this.renderer = renderer;
    }
    /**
     * @return {?}
     */
    FlipComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.rotate();
    };
    /**
     * @param {?} change
     * @return {?}
     */
    FlipComponent.prototype.ngOnChanges = /**
     * @param {?} change
     * @return {?}
     */
    function (change) {
        if (this.flipContainer !== undefined) {
            this.rotate();
        }
    };
    /**
     * @return {?}
     */
    FlipComponent.prototype.rotate = /**
     * @return {?}
     */
    function () {
        if (this.flip) {
            this.renderer.addClass(this.flipContainer.nativeElement, 'rotate');
        }
        else {
            this.renderer.removeClass(this.flipContainer.nativeElement, 'rotate');
        }
    };
    FlipComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngx-flip',
                    template: "<div class=\"flipper\" #flipContainer>\r\n<div class=\"front\">\r\n  <ng-content select=\"[front]\" ></ng-content>\r\n</div>\r\n  <div class=\"back\">\r\n  <ng-content select=\"[back]\" ></ng-content>\r\n  </div>\r\n</div>\r\n\r\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: [":host{-webkit-perspective:1000px;perspective:1000px;display:block}:host .flipper{-webkit-transition:.6s;transition:.6s;-webkit-transform-style:preserve-3d;transform-style:preserve-3d;height:100%;width:100%;position:relative}:host .back,:host .front{width:100%;height:100%;position:absolute;top:0;left:0;-webkit-backface-visibility:hidden;backface-visibility:hidden;-webkit-transition:z-index .3s ease-in-out;transition:z-index .3s ease-in-out}:host .front{z-index:2;-webkit-transform:rotateY(0);transform:rotateY(0)}:host .back{-webkit-transform:rotateY(180deg);transform:rotateY(180deg)}:host .rotate{-webkit-transform:rotateY(180deg);transform:rotateY(180deg)}:host .rotate .front{z-index:1!important}:host .rotate .back{z-index:2!important}"]
                }] }
    ];
    /** @nocollapse */
    FlipComponent.ctorParameters = function () { return [
        { type: Renderer2 }
    ]; };
    FlipComponent.propDecorators = {
        flip: [{ type: Input }],
        flipContainer: [{ type: ViewChild, args: ['flipContainer', { static: true },] }]
    };
    return FlipComponent;
}());
export { FlipComponent };
if (false) {
    /** @type {?} */
    FlipComponent.prototype.flip;
    /**
     * @type {?}
     * @private
     */
    FlipComponent.prototype.flipContainer;
    /**
     * @type {?}
     * @private
     */
    FlipComponent.prototype.renderer;
}
